airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -0,0 +1,129 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, Iterable, Mapping, Optional
+
+
+class CheckpointMode(Enum):
+    INCREMENTAL = "incremental"
+    RESUMABLE_FULL_REFRESH = "resumable_full_refresh"
+    FULL_REFRESH = "full_refresh"
+
+
+class CheckpointReader(ABC):
+    """
+    CheckpointReader manages how to iterate over a stream's partitions and serves as the bridge for interpreting the current state
+    of the stream that should be emitted back to the platform.
+    """
+
+    @abstractmethod
+    def next(self) -> Optional[Mapping[str, Any]]:
+        """
+        Returns the next slice that will be used to fetch the next group of records
+        """
+
+    @abstractmethod
+    def observe(self, new_state: Mapping[str, Any]) -> None:
+        """
+        Updates the internal state of the checkpoint reader based on the incoming stream state from a connector.
+
+        WARNING: This is used to retain backwards compatibility with streams using the legacy get_stream_state() method.
+        In order to uptake Resumable Full Refresh, connectors must migrate streams to use the state setter/getter methods.
+        """
+        # todo blai: Ideally observe and get_checkpoint should just be one method, but because of the legacy state behavior
+        #  observation and reading Stream.state checkpoint are not 1:1 with each other
+
+    @abstractmethod
+    def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
```

Comment:
See the comment a few lines above. Ideally we should simply have one checkpoint method that is called at the end of a slice or after syncing all records of a stream that reads `stream.state` and the checkpoint reader doesn't have to manage its own internal state.

However, we retain the deprecated legacy method `get_updated_state()` which manages state by comparing it against emitted records. The legacy behavior forces us to decouple updating state and when to emit state since we otherwise would emit state too frequently after every record.

Given the checkpoint reader is not a public interface or intended for use outside the CDK, we can tolerate the interface not being perfect, but I did want to highlight why I had to do this

Reply:
It's just that now there's an extra interface method that could have otherwise been consolidated. Not a deal breaker but aside from the legacy case, we were always following up the update call w/ the emit state.